frame/staking-rewards/src/lib.rs

```diff
@@ -493,11 +514,54 @@ pub mod pallet {
 
 		#[transactional]
 		fn unstake(
-			_who: &Self::AccountId,
-			_position: &Self::PositionId,
+			who: &Self::AccountId,
+			position_id: &Self::PositionId,
 			_remove_amount: Self::Balance,
 		) -> DispatchResult {
-			Err("Not implemented".into())
+			let keep_alive = false;
+			let stake = Stakes::<T>::try_get(position_id).map_err(|_| Error::<T>::StakeNotFound)?;
+			ensure!(who == &stake.owner, Error::<T>::OnlyOwnerCanDoIt);
+			let pool_id = stake.reward_pool_id;
+			let mut rewards_pool =
+				RewardPools::<T>::try_get(pool_id).map_err(|_| Error::<T>::RewardsPoolNotFound)?;
+
+			let mut inner_rewards = rewards_pool.rewards.into_inner();
+			for (asset_id, reward) in inner_rewards.iter_mut() {
+				let inflation =
+					stake.reductions.get(asset_id).ok_or(Error::<T>::ReductionConfigProblem)?;
+				let claim = if rewards_pool.total_shares == Zero::zero() {
+					Zero::zero()
+				} else {
+					reward
+						.total_rewards
+						.safe_mul(&stake.share)?
+						.safe_div(&rewards_pool.total_shares)?
+						.safe_sub(inflation)?
+				};
+				reward.total_rewards = reward.total_rewards.safe_sub(inflation)?;
```

Comment:
As we are transferring out the claim we need to reduce also that from the total rewards available

Reply:
I see that this piece is missing from the original algo also, I will update that.

Reply:
Also we need to make sure that the claim we transfer out does not exceed the total_rewards available in the pool.
eg: `claim = min(claim, total_rewards)`

Reply:
Fixed.